dagshub/models/model_locator.py

```diff
@@ -0,0 +1,320 @@
+import logging
+import os
+from dataclasses import dataclass
+from enum import auto, IntEnum
+from functools import cached_property
+from pathlib import Path, PurePosixPath
+from typing import Optional, Literal, Union, Tuple, List
+
+import dacite
+import yaml
+from pathvalidate import sanitize_filepath
+
+from dagshub.common.analytics import send_analytics_event
+from dagshub.common.api import RepoAPI
+from dagshub.common.api.repo import PathNotFoundError
+from dagshub.common.api.responses import StorageAPIEntry
+from dagshub.common.determine_repo import determine_repo
+from dagshub.common.helpers import log_message
+from dagshub.models.model_loaders import ModelLoader, RepoModelLoader, BucketModelLoader, DagsHubStorageModelLoader
+
+# TODO:
+"""
+Ways to get model:
+- automatic - with the cascade
+- from path
+- from bucket
+- from mlflow model
+- from mlflow artifact
+"""
+
+# TODO: don't forget to separate dvc downloads by the hash of the dir
+
+logger = logging.getLogger(__name__)
+
+
+class ModelNotFoundError(Exception):
+    def __str__(self):
+        return "Could not find model"
+
+
+@dataclass
+class ModelYamlStruct:
+    model_dir: str
+
+
+class StorageType(IntEnum):
+    Repo = auto()
+    DagshubStorage = auto()
+    Bucket = auto()
+
+
+class ModelLocator:
+    def __init__(
+        self,
+        repo: Optional[Union[str, RepoAPI]] = None,
+        path: Optional[str] = None,
+        bucket: Optional[str] = None,
+        git_ref: Optional[str] = None,
+        mlflow_model: Optional[str] = None,
+        mlflow_artifact: Optional[str] = None,
+        download_dest: Optional[Union[str, os.PathLike]] = None,
+        download_type: Literal["lazy", "eager"] = "eager",
+    ):
+        self._repo = repo
+        self.path = path
+        self.bucket = bucket
+        self._git_ref = git_ref
+        self.mlflow_model = mlflow_model
+        self.mlflow_artifact = mlflow_artifact
+        self._download_dest = download_dest
+        self.download_type = download_type
+
+    @cached_property
+    def repo_api(self) -> RepoAPI:
+        if isinstance(self._repo, RepoAPI):
+            return self._repo
+        if self._repo is None:
+            repo_api, git_ref = determine_repo()
+            if self._git_ref is None:
+                self._git_ref = git_ref
+            return repo_api
+        return RepoAPI(self._repo)
+
+    @property
+    def git_ref(self):
+        if self._git_ref is None:
+            self._git_ref = self.repo_api.default_branch
+        return self._git_ref
+
+    @cached_property
+    def download_destination(self) -> Path:
+        if self._download_dest is not None:
+            return Path(self._download_dest)
+        return Path(sanitize_filepath(os.path.join(Path.home(), "dagshub", "models", self.repo_api.full_name)))
```

Comment:
Should this be customizable with an env variable? 
Is there already a similar path for dagshub resources? Should this use a common logic with the token cache location for example?

Reply:
I don't want to reuse the token cache location for this, because it uses kind of an "AppData" directory, that isn't supposed to be used for file storage
These models will be potentially very big, so they could grow in size a lot, so I want it to be discoverable by users.
We're already using `~/dagshub` for data engine files, so it is reusing already existing path in a way.

Reply:
But kind of same comment as before, the different local paths aren't really managed anywhere, and they're spread inline, so I was implying that I expected to see some DRY for how the path is built. This would allow later changes to be less painful. Not critical for now anyway.